File: LeetCode/Problem 179.py
# LeetCode: Problem 179
# Jonathan Kosasih

#Given a list of non negative integers, arrange them such that they form the largest number.
#
#Example 1:
#Input: [10,2]
#Output: "210"
#
#Example 2:
#Input: [3,30,34,5,9]
#Output: "9534330"
#Note: The result may be very large, so you need to return a string instead of an integer.
#------------------------------------------------------------------
# STUDY SOLUTION!
# algorithm:
# convert nums to str entries
# sort entries using custom comparator (x+y > y+x)
# join strings, if 0 is first entry just return 0 (case of all 0s inputted)
class LargerNumKey(str):
    def __lt__(x, y):
        return x+y > y+x
        
class Solution(object):
    def largestNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: str
        """
        largest_num = ''.join(sorted(map(str, nums), key=LargerNumKey))
        return '0' if largest_num[0] == '0' else largest_num
        
        
sol = Solution()
example = [10,2]
print(str(sol.largestNumber(example)))
example = [3,30,34,5,9]
print(str(sol.largestNumber(example)))
example = [1,2,3,4,5]
print(str(sol.largestNumber(example)))

File: LeetCode/C++/Problem 179.cpp
// LeetCode: Problem 179
// Jonathan Kosasih
/*
Given a list of non negative integers, arrange them such that they form the largest number.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

class Solution {
	struct comp {
		bool operator() (int a, int b) {
			string comp1 = to_string(a) + to_string(b);
			string comp2 = to_string(b) + to_string(a);
			return comp1 > comp2;
		}
	} myComp;
public:
	string largestNumber(vector<int>& nums) {
		sort(nums.begin(), nums.end(), myComp);
		if (nums[0] == 0) return "0";
		string ans = "";
		for (auto num : nums) {
			ans = ans + to_string(num);
		}
		return ans;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}